### ISSUE TITLE
Can't inject arguments in methods inside a decorator.

### ISSUE THREAD

**Issue by AlexanderHOtt** [microsoft/pyright#3456]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When using a decorator, you can't inject values as arguments into methods because pyright thinks the "self" argument will not be automatically passed.

**To Reproduce**
```py
import typing as t
import collections_abc as c
P = t.ParamSpec("P")
R = t.TypeVar("R")

def inj_int(func: c.Callable[t.Concatenate["Foo", int, P], R]) -> c.Callable[t.Concatenate["Foo", P], R]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        return func(1, *args, **kwargs)

    return wrapper  # Pyright: Expression of type "(*args: P.args, **kwargs: P.kwargs) -> R@inj_int" cannot be assigned
    # to return type "(Foo, **P@inj_int) -> R@inj_int"

class Foo:
    @inj_int
    def f(self, x: int, y: str) -> str:
        return f"{x}{y}"

f = Foo()
f.f("a")
```

**Expected behavior**
I can inject arguments into the method.

**VS Code extension or command-line**
CLI: `pyright 1.1.246`

**Comment by AlexanderHOtt** [comment]
After further inspection of my code, it looks like python injects the `1` before the `self`, argument, so the correct method signature would be
```py
class Foo:
    def f(x: int, self, y:str) -> None:
        print(f"{self!r}\n{x!r}\n{y!r}")
```
which is not good. Is there any way to inject arguments into methods at all?

**Comment by erictraut** [comment]
```python
import typing as t
import collections.abc as c

P = t.ParamSpec("P")
R = t.TypeVar("R")
TFoo = t.TypeVar("TFoo", bound="Foo")

def inj_int(
    func: c.Callable[t.Concatenate[TFoo, int, P], R]
) -> c.Callable[t.Concatenate[TFoo, P], R]:
    def wrapper(_self: TFoo, *args: P.args, **kwargs: P.kwargs) -> R:
        return func(_self, 1, *args, **kwargs)

    return wrapper

class Foo:
    @inj_int
    def f(self, x: int, y: str) -> str:
        print(x, y)
        return f"{x}{y}"

f = Foo()
f.f("a")
```

**Comment by AlexanderHOtt** [comment]
Thanks, I didn't know you could manipulate the `self` argument like that.
